src/Bootstrap/Carousel.elm

```diff
@@ -0,0 +1,509 @@
+            -- don't trigger new animations when already animating
+            Sub.none
+
+
+
+--- Update ---
+
+
+type Msg
+    = Cycle
+    | Pause
+    | StartTransition Transition
+    | SetAnimating
+    | EndTransition
+    | SetHover Bool
+
+
+update : Msg -> State -> State
+update message ((State tstage ({ currentIndex, hovered } as settings)) as model) =
+    case message of
+        Pause ->
+            State tstage { settings | running = Paused }
+
+        Cycle ->
+            State tstage { settings | running = Active }
+
+        SetHover isHovered ->
+            State tstage { settings | hovered = Maybe.map (\_ -> isHovered) hovered }
+
+        StartTransition transition ->
+            let
+                newSettings =
+                    case settings.running of
+                        WaitForUser ->
+                            -- the user has clicked something, so we can now automatically cycle
+                            { settings | running = Active }
+
+                        _ ->
+                            settings
+            in
+                case tstage of
+                    NotAnimating ->
+                        if nextIndex (Start transition) currentIndex /= currentIndex then
+                            State (Start transition) newSettings
+                        else
+                            -- don't do anything if animating to the current index
+                            State tstage newSettings
+
+                    _ ->
+                        -- don't start another animation when one is running
+                        State tstage newSettings
+
+        SetAnimating ->
+            case tstage of
+                NotAnimating ->
+                    model
+
+                Start transition ->
+                    State (Animating transition) settings
+
+                Animating transition ->
+                    State (Animating transition) settings
+
+        EndTransition ->
+            case tstage of
+                NotAnimating ->
+                    -- should never happen
+                    model
+
+                _ ->
+                    State NotAnimating { settings | currentIndex = nextIndex tstage currentIndex }
+
+
+{-| Move the carousel to the next slide.
+
+When the transition is invalid, nothing will happen.
+-}
+next : State -> State
+next =
+    update (StartTransition Next)
+
+
+{-| Move the carousel to the previous slide.
+
+When the transition is invalid, nothing will happen.
+-}
+prev : State -> State
+prev =
+    update (StartTransition Prev)
+
+
+{-| Move the carousel to the nth slide
+
+When the transition is invalid, nothing will happen.
+-}
+toSlide : Int -> State -> State
+toSlide n =
+    update (StartTransition (Number n))
+
+
+{-| Calculate the next index based on the current state
+
+This is hardcoded to cycle between the first 3 slides for now
+-}
+nextIndex : TransitionStage Transition -> Int -> Int
```

Comment:
This is a tricky one. Need to think more on it, but can't think of a nice solution to this at the moment.

The upside of the toMsg updates state approach is that from the view you always also know how many slides you have. However for the subscription the problem still remains I guess.